src/crimson/os/seastore/cache.cc

```diff
@@ -1296,27 +1296,36 @@ void Cache::backref_batch_update(
   LOG_PREFIX(Cache::backref_batch_update);
   DEBUG("inserting {} entries", list.size());
   if (!backref_buffer) {
-    backref_buffer = std::make_unique<backref_buffer_t>();
+    backref_buffer = std::make_unique<backref_cache_t>();
   }
   // backref_buf_entry_t::laddr == L_ADDR_NULL means erase
   for (auto &ent : list) {
     if (ent->laddr == L_ADDR_NULL) {
-      auto [it, insert] = backref_remove_set.insert(*ent);
-      boost::ignore_unused(insert);
+      auto insert_set_iter = backref_inserted_set.find(
+	ent->paddr, backref_buf_entry_t::cmp_t());
+      if (insert_set_iter == backref_inserted_set.end()) {
+	// backref to be removed isn't in the backref buffer,
+	// it must be in the backref tree.
+	auto [it, insert] = backref_remove_set.insert(*ent);
+	boost::ignore_unused(insert);
 #ifndef NDEBUG
-      if (!insert) {
-	ERROR("backref_remove_set already contains {}", ent->paddr);
-      }
+	if (!insert) {
+	  ERROR("backref_remove_set already contains {}", ent->paddr);
+	}
 #endif
-      assert(insert);
-    } else {
-#ifndef NDEBUG
-      auto r = backref_remove_set.erase(ent->paddr, backref_buf_entry_t::cmp_t());
-      if (r) {
-	ERROR("backref_remove_set contains: {}", ent->paddr);
+	assert(insert);
+      } else {
+	// the backref insertion hasn't been applied to the
+	// backref tree
+	auto seq = insert_set_iter->seq;
+	auto it = backref_buffer->backrefs.find(seq);
+	ceph_assert(it != backref_buffer->backrefs.end());
+	auto &backref_buf = it->second;
+	backref_buf.br_list.erase(
+	  backref_buf_entry_t::list_t::s_iterator_to(*insert_set_iter));
```

Comment:
does the entry also need to be removed from `backref_buf.backrefs`?

or does this mean `backref_buf_t::backrefs` may not be consistent with `backref_buf_t::br_list`?

also assert backref_buf.br_list.erase() should be always successful.

Reply:
> does this mean `backref_buf_t::backrefs` may not be consistent with `backref_buf_t::br_list`?

Yes